File: frontend/src/app/services/policy.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Policy } from '../models/policy.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class PolicyService {
  private apiUrl = `${environment.apiUrl}/policies`;

  constructor(private http: HttpClient) {}

  getAllPolicies(): Observable<Policy[]> {
    return this.http.get<Policy[]>(this.apiUrl);
  }

  getPolicyById(id: number): Observable<Policy> {
    return this.http.get<Policy>(`${this.apiUrl}/${id}`);
  }

  getPoliciesByCustomer(customerId: number): Observable<Policy[]> {
    return this.http.get<Policy[]>(`${this.apiUrl}/customer/${customerId}`);
  }

  getPoliciesByType(policyType: string): Observable<Policy[]> {
    return this.http.get<Policy[]>(`${this.apiUrl}/type/${policyType}`);
  }

  createMotorPolicy(data: any): Observable<Policy> {
    return this.http.post<Policy>(`${this.apiUrl}/motor`, data);
  }

  createEndowmentPolicy(data: any): Observable<Policy> {
    return this.http.post<Policy>(`${this.apiUrl}/endowment`, data);
  }

  createHousePolicy(data: any): Observable<Policy> {
    return this.http.post<Policy>(`${this.apiUrl}/house`, data);
  }

  createCommercialPolicy(data: any): Observable<Policy> {
    return this.http.post<Policy>(`${this.apiUrl}/commercial`, data);
  }

  updatePolicy(id: number, policy: Policy): Observable<Policy> {
    return this.http.put<Policy>(`${this.apiUrl}/${id}`, policy);
  }

  deletePolicy(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

// Made with Bob

File: backend/src/main/java/com/ibm/genapp/service/PolicyService.java
package com.ibm.genapp.service;

import com.ibm.genapp.exception.ResourceNotFoundException;
import com.ibm.genapp.model.*;
import com.ibm.genapp.repository.CustomerRepository;
import com.ibm.genapp.repository.PolicyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service layer for Policy business logic
 */
@Service
@RequiredArgsConstructor
@Transactional
public class PolicyService {

    private final PolicyRepository policyRepository;
    private final CustomerRepository customerRepository;

    public List<Policy> getAllPolicies() {
        return policyRepository.findAll();
    }

    public Policy getPolicyById(Long id) {
        return policyRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Policy", "id", id));
    }

    public List<Policy> getPoliciesByCustomerId(Long customerId) {
        return policyRepository.findByCustomerCustomerNumber(customerId);
    }

    public List<Policy> getPoliciesByType(String policyType) {
        return policyRepository.findByPolicyType(policyType);
    }

    public Policy createMotorPolicy(Long customerId, Policy policy, Motor motor) {
        Customer customer = customerRepository.findById(customerId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", customerId));
        
        policy.setCustomer(customer);
        policy.setPolicyType("M");
        Policy savedPolicy = policyRepository.save(policy);
        
        motor.setPolicy(savedPolicy);
        motor.setPolicyNumber(savedPolicy.getPolicyNumber());
        savedPolicy.setMotor(motor);
        
        return policyRepository.save(savedPolicy);
    }

    public Policy createEndowmentPolicy(Long customerId, Policy policy, Endowment endowment) {
        Customer customer = customerRepository.findById(customerId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", customerId));
        
        policy.setCustomer(customer);
        policy.setPolicyType("E");
        Policy savedPolicy = policyRepository.save(policy);
        
        endowment.setPolicy(savedPolicy);
        endowment.setPolicyNumber(savedPolicy.getPolicyNumber());
        savedPolicy.setEndowment(endowment);
        
        return policyRepository.save(savedPolicy);
    }

    public Policy createHousePolicy(Long customerId, Policy policy, House house) {
        Customer customer = customerRepository.findById(customerId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", customerId));
        
        policy.setCustomer(customer);
        policy.setPolicyType("H");
        Policy savedPolicy = policyRepository.save(policy);
        
        house.setPolicy(savedPolicy);
        house.setPolicyNumber(savedPolicy.getPolicyNumber());
        savedPolicy.setHouse(house);
        
        return policyRepository.save(savedPolicy);
    }

    public Policy createCommercialPolicy(Long customerId, Policy policy, Commercial commercial) {
        Customer customer = customerRepository.findById(customerId)
            .orElseThrow(() -> new ResourceNotFoundException("Customer", "id", customerId));
        
        policy.setCustomer(customer);
        policy.setPolicyType("C");
        Policy savedPolicy = policyRepository.save(policy);
        
        commercial.setPolicy(savedPolicy);
        commercial.setPolicyNumber(savedPolicy.getPolicyNumber());
        savedPolicy.setCommercial(commercial);
        
        return policyRepository.save(savedPolicy);
    }

    public Policy updatePolicy(Long id, Policy policyDetails) {
        Policy policy = getPolicyById(id);
        
        policy.setIssueDate(policyDetails.getIssueDate());
        policy.setExpiryDate(policyDetails.getExpiryDate());
        policy.setBrokerId(policyDetails.getBrokerId());
        policy.setBrokersReference(policyDetails.getBrokersReference());
        policy.setPayment(policyDetails.getPayment());
        policy.setCommission(policyDetails.getCommission());
        
        return policyRepository.save(policy);
    }

    public void deletePolicy(Long id) {
        Policy policy = getPolicyById(id);
        policyRepository.delete(policy);
    }
}

// Made with Bob
